src/filtering.jl

```diff
@@ -0,0 +1,133 @@
+# Use a custom struct so accessing a non-existent field (i.e. trying to filer on anything
+# other than `name` or `tags`) gives a somewhat informative error.
+struct TestItemMetadata
+    name::String
+    tags::Vector{Symbol}
+end
+
+struct TestItemFilter{
+        F<:Function,
+        T<:Union{Nothing,Symbol,AbstractVector{Symbol}},
+        N<:Union{Nothing,AbstractString,Regex}
+} <: Function
+    shouldrun::F
+    tags::T
+    name::N
+    strict::Bool  # TODO: hardcode `strict=true`
+end
+
+function (f::TestItemFilter)(ti::TestItemMetadata)
+    return f.shouldrun(ti)::Bool && _shouldrun(f.tags, ti) && _shouldrun(f.name, ti)
+end
+# TODO: drop this method with we hardcode `strict=true`
+function (f::TestItemFilter)(ti::TestItem)
+    if f.strict # we were able to do all filtering already on the AST
+        return true
+    else
+        return f.shouldrun(ti)::Bool && _shouldrun(f.tags, ti) && _shouldrun(f.name, ti)
+    end
+end
+
+_shouldrun(name::AbstractString, ti) = name == ti.name
+_shouldrun(pattern::Regex, ti) = contains(ti.name, pattern)
+_shouldrun(tags::AbstractVector{Symbol}, ti) = issubset(tags, ti.tags)
+_shouldrun(tag::Symbol, ti) = tag in ti.tags
+_shouldrun(::Nothing, ti) = true
+
+# Filter the AST before the code is evaluated.
+# Called when a `TestItemFilter` is passed as the first argument to `Base.include`.
+function (f::TestItemFilter)(expr::Expr)
+    if Meta.isexpr(expr, :error)
+        # If the expression failed to parse, most user-friendly to throw the ParseError,
+        # rather than report an error about using only `@testitem` or `@testsetup`.
+        Core.eval(Main, expr)
+    end
+    is_retestitem_macrocall(expr, f.strict) || _throw_not_macrocall(expr)
+    expr = filter_testitem(f, expr)
+    return expr
+end
+
+# Filter out `@testitem` calls based on the `name` and `tags` keyword passed to `runtests`.
+# Any other macro calls (e.g. `@testsetup`) are left as is.
+# If the `@testitem` call is not as expected, it is left as is so that it throws an error.
+#
+# Replacing the expression with `nothing` effectively removes it from the file.
+# `Base.include` will still call `Core.eval(M, nothing)` which has a tiny overhead,
+# but less than `Core.eval(M, :())`. We could instead replace `Base.include` with a
+# custom function that doesn't call `Core.eval(M, expr)` if `expr === nothing`, but
+# using `Base.include` means we benefit from improvements made upstream rather than
+# having to maintain our own version of that code.
+function filter_testitem(f, expr)
```

Comment:
I got confused a little because there are two points at which we filter test items -- here at the AST level and then later in the `FilteredVector` level when we filter on the `eval`d `TestItem`. Let's maybe explain here, the things that need to be expanded, like:
```julia
@testitem "name with interpolation $(VERSION)" begin ... end
```
get filtered later and where? Because I think I'll forget how things work again:)

Reply:
ooo, name with interpolation is interesting... i've never thought of us as (officially) supporting that.

i always thought we'd (one day) require files only have `@testitem` and `@testsetup` calls at the top-level, and that `name` and `tags` must be a static

But yes, you're quite right, for things that need to be expanded before filtering (`@test_rel`, `@eval ...`) we have to filter later in the `FilteredVector`, which is why (one day) i'd like us to not support those things at all and do all filter on the AST. Will add more comments explaining this.